ui/src/QrScan.tsx

```diff
@@ -0,0 +1,259 @@
+import { Button, Grid, Box } from '@mui/material';
+import { useTheme } from '@mui/material/styles';
+import { styled } from '@mui/system';
+import { useNavigate, useLocation } from 'react-router-dom';
+import { v4 as uuidv4 } from 'uuid';
+import frame from './assets/frame.png';
+import { useQrDataContext } from './QrDataProvider';
+import QrScanner from 'qr-scanner';
+import { useErrorHandler } from 'react-error-boundary';
+
+let qrScan: QrScanner;
+
+const QrScan = () => {
+  const navigate = useNavigate();
+  const location = useLocation();
+  const handleErrorFallback = useErrorHandler();
+  const { setQrCodes, resetQrCodes, qrCodes } = useQrDataContext();
+  const [scannedCodes, setScannedCodes] = useState<(null | string)[]>([]);
+  const [scannedData, setScannedData] = useState<string>('');
+  const runningQrScanner = useRef<null | QrScanner>(null);
+  const scannedCodesRef = useRef<(null | string)[]>([]);
+
+  const theme = useTheme();
+  const classes = {
+    button: {
+      '&:hover': {
+        cursor: 'default',
+      },
+    },
+    box: {
+      display: 'flex',
+      alignItems: 'center',
+      justifyContent: 'center',
+      width: '100%',
+      marginTop: '2em',
+    },
+    grid: {
+      display: 'flex',
+      flexDirection: 'column',
+      flexWrap: 'nowrap',
+      alignItems: 'center',
+      justifyContent: 'center',
+      width: 'auto',
+      [theme.breakpoints.down('md')]: {
+        maxHeight: '550px',
+        maxWidth: '300px',
+        margin: '1rem',
+      },
+      [theme.breakpoints.up('md')]: {
+        maxHeight: '550px',
+        maxWidth: '650px',
+        margin: '2rem',
+      },
+    },
+    gridContainerMultiple: {
+      display: 'flex',
+      flexDirection: 'row',
+      flexGrow: 1,
+      alignItems: 'right',
+      justifyContent: 'right',
+      paddingBottom: '1rem',
+    },
+    gridItem: {
+      display: 'flex',
+      position: 'relative',
+      alignItems: 'center',
+      justifyContent: 'center',
+      width: '100%',
+      height: '100%',
+    },
+  };
+
+  const StyledImg = styled('img')({
+    position: 'relative',
+    [theme.breakpoints.down('md')]: {
+      maxHeight: '550px',
+      maxWidth: '300px',
+    },
+    [theme.breakpoints.up('md')]: {
+      maxHeight: '550px',
+      maxWidth: '650px',
+    },
+    objectFit: 'contain',
+    zIndex: '2',
+  });
+
+  const confirmSHLCreation = () => {
+    if (window.confirm('SMART Health Card successfully scanned. Create new SMART Health Link?') === true) {
+      return true;
+    } else {
+      return false;
+    }
+  };
+
+  const handleError = useCallback(() => {
+    navigate('/error');
+  }, [navigate]);
+
+  useEffect(() => () => {
+    if (runningQrScanner.current) {
+      runningQrScanner.current.stop();
+    }
+  }, [])
+
+
+  const videoRef = useRef<any>(null);
+  // Get user media when the page first renders, and feed into createQrScanner()
+  useEffect(() => {
+    const getUserMedia = async () => {
```

Comment:
Can these two inner `const` functions be lifted out of `useEffect` and just called from the effect? It's a bit hard to follow what the effect is doing specifically. It's also good to avoid re-defining each time the effect runs.